## Atsu Atakpa
## Homework 1

1.	Install Python (if you don’t have it already), and install NLTK.  

2.	Follow the instructions in chapter 1 of Bird-Klein for implementing a “lexical diversity” scoring routine.

3.	Go to http://www.gutenberg.org/wiki/Children%27s_Instructional_Books_(Bookshelf), and obtain three texts (of different grade levels) from the “Graded Readers” section. Report the lexical diversity score of each. Explain whether the result was surprising.

4.	Also compare the vocabulary size of the same three texts. Explain whether the result was surprising.  

5.	Write a paragraph arguing whether vocabulary size and lexical diversity in combination could be a better measure of text difficulty (or reading level) than either measure is by itself.


In [1]:
import sys
sys.version
reload(sys)
sys.setdefaultencoding('utf-8')


In [2]:
from __future__ import division
import nltk
#from nltk import word_tokenize
from nltk import word_tokenize,sent_tokenize
print('The nltk version is {}.'.format(nltk.__version__))

2. Follow the instructions in chapter 1 of Bird-Klein for implementing a “lexical diversity” scoring routine.

In [3]:
def lexical_diversity(text):
    return len(set(text)) / len(text)

In [4]:
def percentage(count, total):
    return 100 * count / total

Book One

In [5]:
first_text = open("C:/Users/atsua/OneDrive/Documents/SMU/7. Natural Language Processing/National_First_Reader.txt","r")

In [6]:
first_text = first_text.read()

In [7]:
first_text_tokenize = nltk.word_tokenize(first_text)

In [8]:
sorted(set(first_text_tokenize))

['!',
 '#',
 '$',
 '%',
 '&',
 "'",
 "''",
 "'AS-IS",
 "'S",
 "'ll",
 "'s",
 '(',
 ')',
 '*',
 '***',
 '*****',
 ',',
 '-',
 '--',
 '.',
 '//gutenberg.net/license',
 '//pglaf.org',
 '//pglaf.org/donate',
 '//pglaf.org/fundraising',
 '//www.gutenberg.net',
 '//www.gutenberg.net/1/3/8/5/13853/',
 '//www.pglaf.org',
 '0',
 '1',
 '1.A',
 '1.B',
 '1.C',
 '1.D',
 '1.E',
 '1.E.1',
 '1.E.2',
 '1.E.3',
 '1.E.4',
 '1.E.5',
 '1.E.6',
 '1.E.7',
 '1.E.8',
 '1.E.9',
 '1.F',
 '1.F.1',
 '1.F.2',
 '1.F.3',
 '1.F.4',
 '1.F.5',
 '1.F.6',
 '10th.',
 '11th.',
 '12th.',
 '13853',
 '13853-0.txt',
 '13853-0.zip',
 '13th.',
 '1500',
 '1888',
 '1st.',
 '2',
 '20',
 '2001',
 '2004',
 '25',
 '2d.',
 '3',
 '30',
 '36',
 '3d.',
 '4',
 '44',
 '4557',
 '4th.',
 '5',
 '5,000',
 '50',
 '501',
 '52',
 '596-1887',
 '5th.',
 '6',
 '60',
 '64-6221541',
 '68',
 '6th.',
 '7',
 '7th.',
 '8',
 '801',
 '809',
 '84116',
 '8th.',
 '9',
 '90',
 '99712.',
 '9th.',
 ':',
 ';',
 '?',
 '@',
 'A',
 'A.S.',
 'ACTUAL',
 'AGREE',
 'AGREEM

In [9]:
len(first_text_tokenize)

13306

In [10]:
len(set(first_text_tokenize))

2201

In [11]:
len(set(word.lower() for word in first_text_tokenize)) # Number of unique tokens when capitalization is dealt with.

1928

In [12]:
len(set(word.lower() for word in first_text_tokenize if word.isalpha())) # Number of unique tokens when capitalization is dealt with.

1274

In [13]:
sorted_words=sorted((set(word.lower() for word in first_text_tokenize if word.isalpha())))

In [14]:
lexical_diversity(first_text_tokenize) # this is th score for the version that has not been cleaned up for captiliztion and puncuation

0.16541409890275063

In [15]:
sorted_words

['a',
 'abide',
 'ability',
 'about',
 'above',
 'absolutely',
 'abstract',
 'accept',
 'accepted',
 'accepting',
 'access',
 'accessed',
 'accessible',
 'accordance',
 'accurate',
 'actions',
 'active',
 'actual',
 'adaptable',
 'adapted',
 'addition',
 'additional',
 'additions',
 'address',
 'addresses',
 'advancement',
 'afraid',
 'after',
 'again',
 'against',
 'agent',
 'ago',
 'agree',
 'agreed',
 'agreement',
 'ak',
 'al',
 'all',
 'allow',
 'almost',
 'alone',
 'alphabets',
 'already',
 'also',
 'alteration',
 'alternate',
 'always',
 'am',
 'american',
 'among',
 'an',
 'and',
 'anger',
 'ann',
 'another',
 'answer',
 'any',
 'anyone',
 'anything',
 'anywhere',
 'appear',
 'appearing',
 'appears',
 'apple',
 'apples',
 'applicable',
 'apply',
 'approach',
 'archive',
 'are',
 'arise',
 'arrangement',
 'array',
 'artificial',
 'artists',
 'as',
 'ascii',
 'ask',
 'assistance',
 'associated',
 'at',
 'attached',
 'attention',
 'author',
 'authors',
 'available',
 'average',
 'a

In [16]:
lexical_diversity(sorted_words)

1.0

Book two

In [17]:
second_text = open("C:/Users/atsua/OneDrive/Documents/SMU/7. Natural Language Processing/Animal_Heroes.txt","r")

In [18]:
second_text = second_text.read()
second_text_tokenize = nltk.word_tokenize(second_text)

In [19]:
sorted(set(second_text_tokenize))

['!',
 '#',
 '$',
 '%',
 "'",
 "''",
 "'92",
 "'AS-IS",
 "'Badlands",
 "'Cottonwood",
 "'Henglish-man",
 "'It",
 "'Jack",
 "'Lost",
 "'Royal",
 "'T",
 "'Yes",
 "'all-powerful",
 "'as",
 "'ave",
 "'barkeep",
 "'breeding",
 "'d",
 "'daily",
 "'daily'",
 "'death",
 "'em",
 "'er",
 "'four",
 "'left",
 "'lighting",
 "'ll",
 "'m",
 "'no-breakfast",
 "'ole",
 "'orses",
 "'ow",
 "'owld",
 "'pearl",
 "'re",
 "'red-snow",
 "'s",
 "'sniff",
 "'soon",
 "'spec",
 "'sporting",
 "'tater-bake",
 "'tind",
 "'ve",
 "'ye",
 '(',
 ')',
 '***',
 '*****',
 ',',
 '-',
 '--',
 '-a',
 '.',
 '//gutenberg.net/license',
 '//pglaf.org',
 '//pglaf.org/donate',
 '//pglaf.org/fundraising',
 '//www.gutenberg.net',
 '//www.gutenberg.org/2/2/8/2284/',
 '//www.pglaf.org',
 '1',
 '1.A',
 '1.B',
 '1.C',
 '1.D',
 '1.E',
 '1.E.1',
 '1.E.2',
 '1.E.3',
 '1.E.4',
 '1.E.5',
 '1.E.6',
 '1.E.7',
 '1.E.8',
 '1.E.9',
 '1.F',
 '1.F.1',
 '1.F.2',
 '1.F.3',
 '1.F.4',
 '1.F.5',
 '1.F.6',
 '10',
 '12',
 '125th',
 '12:30',
 '1500',
 '1696

In [20]:
len(second_text_tokenize)

68925

In [21]:
len(set(second_text_tokenize))

7921

In [22]:
len(set(word.lower() for word in second_text_tokenize)) # Number of unique tokens when capitalization is dealt with.

7318

In [23]:
len(set(word.lower() for word in second_text_tokenize if word.isalpha())) # Number of unique tokens when capitalization is dealt with.

6552

In [24]:
second_sorted_words=sorted((set(word.lower() for word in second_text_tokenize if word.isalpha())))

In [25]:
lexical_diversity(second_text_tokenize) # this is the score for the version that has not been cleaned up for captiliztion and puncuation

0.11492201668480233

In [26]:
lexical_diversity(second_sorted_words)

1.0

In [27]:
second_sorted_words

['a',
 'abandoned',
 'abandoning',
 'abate',
 'abide',
 'ablaze',
 'able',
 'abode',
 'abounded',
 'about',
 'above',
 'abroad',
 'abrupt',
 'absence',
 'absent',
 'absolute',
 'absolutely',
 'absurdly',
 'abundance',
 'accept',
 'accepted',
 'accepting',
 'access',
 'accessed',
 'accessible',
 'accident',
 'accidental',
 'accidents',
 'accordance',
 'according',
 'account',
 'accustomed',
 'achieved',
 'achievements',
 'acquaintance',
 'acquired',
 'across',
 'act',
 'acted',
 'acting',
 'action',
 'actions',
 'active',
 'actual',
 'actually',
 'acute',
 'add',
 'added',
 'adding',
 'addition',
 'additional',
 'additions',
 'address',
 'addressed',
 'addresses',
 'adjoining',
 'admiration',
 'admit',
 'admitted',
 'admonitions',
 'adopted',
 'adornments',
 'adrift',
 'adroitly',
 'advance',
 'advanced',
 'advantage',
 'advantageously',
 'advent',
 'adventure',
 'adventures',
 'adversary',
 'advisers',
 'afar',
 'affair',
 'affection',
 'affections',
 'affinity',
 'afford',
 'afforded'

In [28]:
type(second_sorted_words)

list

In [29]:
type(second_text_tokenize)

list

#### Question 3
The results are not far from what I was expecting. The text size does impact the lexical diversity score.
#### Question 4
As expected, when we clean the text, we expect the text size to reduce. For example the first book, the number of words from the raw text were 13306 and after cleaning it dropped down to 1274.

#### Question 5
Both the vocabulary size and the lexical diversity is a great measure of text difficulty. However, I'm not sure that should be the main metrics to use to measure text difficulty. I will argure that both text size and lexical diversity should be used to understand the overall structure of our document and be used as a guide and not a final measure.